src/ILCompiler.ReadyToRun/src/ObjectWriter/SectionBuilder.cs

```diff
@@ -0,0 +1,921 @@
+        /// 32-bit relative address from byte following reloc
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_REL32 = 0x10;
+
+        /// <summary>
+        /// Thumb2: based B, BL
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_THUMB_BRANCH24 = 0x13;
+
+        /// <summary>
+        /// Arm64: B, BL
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_BRANCH26 = 0x14;
+
+        /// <summary>
+        /// 32-bit relative address from byte starting reloc
+        /// This is a special NGEN-specific relocation type
+        /// for relative pointer (used to make NGen relocation
+        /// section smaller)
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_RELPTR32 = 0x7C;
+
+        /// <summary>
+        /// 32 bit offset from base of section containing target
+        /// </summary>
+        public const ushort IMAGE_REL_SECREL = 0x80;
+
+        /// <summary>
+        /// ADRP
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_PAGEBASE_REL21 = 0x81;
+
+        /// <summary>
+        /// ADD/ADDS (immediate) with zero shift, for page offset
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_PAGEOFFSET_12A = 0x82;
+
+        /// <summary>
+        /// LDR (indexed, unsigned immediate), for page offset
+        /// </summary>
+        public const ushort IMAGE_REL_BASED_ARM64_PAGEOFFSET_12L = 0x83;
+    }
+
+    /// <summary>
+    /// Opaque key representing a single block.
+    /// </summary>
+    public struct BlockHandle : IEquatable<BlockHandle>, IEqualityComparer<BlockHandle>
+    {
+        /// <summary>
+        /// Internal block handle.
+        /// </summary>
+        private int _value;
+
+        /// <summary>
+        /// Construct the block handle for a given key
+        /// </summary>
+        /// <param name="value">Key value to store in the block handle</param>
+        public BlockHandle(int value)
+        {
+            _value = value;
+        }
+
+        /// <summary>
+        /// True when the block handle is empty.
+        /// </summary>
+        public bool IsNull { get => (_value == default(int)); }
+
+        /// <summary>
+        /// Retrieve value of the block handle
+        /// </summary>
+        /// <returns></returns>
+        public int GetValue()
+        {
+            return _value;
+        }
+
+        bool IEquatable<BlockHandle>.Equals(BlockHandle handle)
+        {
+            return _value == handle._value;
+        }
+
+        bool IEqualityComparer<BlockHandle>.Equals(BlockHandle handle1, BlockHandle handle2)
+        {
+            return handle1._value == handle2._value;
+        }
+
+        int IEqualityComparer<BlockHandle>.GetHashCode(BlockHandle handle)
+        {
+            return handle._value.GetHashCode();
+        }
+    }
+
+    /// <summary>
+    /// A single relocation entry within a block.
+    /// </summary>
+    public struct Relocation
```

Comment:
Are we going to be translating between the concepts that the ILCompiler works with (`ISymbolNode`, `ISymbolDefinitionNode`, `ILCompiler.DependencyAnalysis.Relocation`, `ObjectData`, etc.) and these new types? The types I'm listing are the abstractions between code generators, the compiler, and the object writers, and they're used as such in all of our object writers (the LLVM based one that CoreRT/RyuJIT uses, UTC's object writer we use on the closed source side, or the one for the WebAssembly backend).

Would it be possible to emit from these directly? Or why do we need new types?

Reply:
I agree that ObjectData and DependencyAnalysis.Relocation are quite close to my Block and Relocation objects. My intent was to make the initial implementation mostly self-contained because I don't yet understand all details of the expected interplay between the R2R builder, ILCompiler and JIT but I was trying to get something to work at the same time. We can now easily compare the classes and discuss whether / how to unify them.